java-connectors/kafka-connect-azure-servicebus/src/test/java/io/lenses/streamreactor/connect/azure/servicebus/source/TaskToReceiverBridgeTest.java

```diff
@@ -0,0 +1,130 @@
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ExecutorService;
+import org.apache.kafka.connect.source.SourceRecord;
+import org.apache.kafka.connect.storage.OffsetStorageReader;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+class TaskToReceiverBridgeTest {
+
+  Map<String, String> properties = new HashMap<>();
+  ArrayBlockingQueue<ServiceBusMessageHolder> blockingQueue;
+  OffsetStorageReader offsetReader;
+
+  TaskToReceiverBridge testObj;
+
+  @BeforeEach
+  void setUp() {
+    blockingQueue = mock(ArrayBlockingQueue.class);
+    offsetReader = mock(OffsetStorageReader.class);
+  }
+
+  @Test
+  void closeReceiversShouldCloseAllReceivers() {
+    //given
+    ExecutorService executorService = mock(ExecutorService.class);
+    String receiverId1 = "RECEIVER1";
+    String receiverId2 = "RECEIVER2";
+    ServiceBusReceiverFacade receiver1 = mock(ServiceBusReceiverFacade.class);
+    ServiceBusReceiverFacade receiver2 = mock(ServiceBusReceiverFacade.class);
+    when(receiver1.getReceiverId()).thenReturn(receiverId1);
+    when(receiver2.getReceiverId()).thenReturn(receiverId2);
+    Map<String, ServiceBusReceiverFacade> receivers =
+        Map.of(
+            receiverId1, receiver1,
+            receiverId2, receiver2
+        );
+
+    //when
+    testObj = new TaskToReceiverBridge(blockingQueue, offsetReader, receivers, executorService);
+    testObj.closeReceivers();
+
+    //then
+    verify(receiver1).unsubscribeAndClose();
+    verify(receiver2).unsubscribeAndClose();
+  }
+
+  @Test
+  void pollShouldDrainAllMessagesFromQueue() throws InterruptedException {
+    //given
+    int arrayBlockingQueueCapacity = 10;
+    String receiverId1 = "RECEIVER1";
+    String messageIdTemplate = "MSGID%s";
+    ExecutorService executorService = mock(ExecutorService.class);
+    ServiceBusReceiverFacade receiver1 = mock(ServiceBusReceiverFacade.class);
+    List<SourceRecord> allSourceRecords = new ArrayList<>(arrayBlockingQueueCapacity);
+    BlockingQueue<ServiceBusMessageHolder> sourceRecordBlockingQueue =
+        new ArrayBlockingQueue<>(arrayBlockingQueueCapacity);
+    Map<String, ServiceBusReceiverFacade> receivers = Map.of(receiverId1, receiver1);
+
+    for (int i = 0; i < arrayBlockingQueueCapacity; i++) {
+      ServiceBusMessageHolder mockedRecord = mock(ServiceBusMessageHolder.class);
+      ServiceBusReceivedMessage busReceivedMessage = mock(ServiceBusReceivedMessage.class);
+      SourceRecord sourceRecord = mock(SourceRecord.class);
+      when(mockedRecord.getTranslatedRecord()).thenReturn(sourceRecord);
+      when(mockedRecord.getOriginalRecord()).thenReturn(busReceivedMessage);
+      when(busReceivedMessage.getMessageId()).thenReturn(String.format(messageIdTemplate, i));
+      allSourceRecords.add(sourceRecord);
+      sourceRecordBlockingQueue.put(mockedRecord);
+    }
```

Comment:
A good rule of thumb when working with mocks, is to try and keep the mock creation and behaviour together.

From your code I can tweak it a little to group the mock definition and behaviours and make it a little easier to read:

Also using an IntStream and then a collector can be a good more functional alternative to building an ArrayList incrementally.

```

    List<SourceRecord> allSourceRecords = IntStream.range(0, arrayBlockingQueueCapacity).mapToObj(i -> {

      SourceRecord sourceRecord = mock(SourceRecord.class);

      ServiceBusReceivedMessage busReceivedMessage = mock(ServiceBusReceivedMessage.class);
      when(busReceivedMessage.getMessageId()).thenReturn(String.format(messageIdTemplate, i));

      ServiceBusMessageHolder mockedRecord = mock(ServiceBusMessageHolder.class);
      when(mockedRecord.getTranslatedRecord()).thenReturn(sourceRecord);
      when(mockedRecord.getOriginalRecord()).thenReturn(busReceivedMessage);
      try {
        sourceRecordBlockingQueue.put(mockedRecord);
      } catch (InterruptedException e) {
        throw new RuntimeException(e);
      }
      return sourceRecord;
    }).collect(Collectors.toList());
```

From this point you could start extracting out methods for reusability/clarity.

Reply:
I can get along on grouping, using Stream or for-loop gives the same results.

I wonder however, why try-catch here instead of just telling that test can throw Exception? It's not a production code.

Reply:
@GoMati-MU this is from [your code.](
https://github.com/lensesio/stream-reactor/blob/634953a5877c38377ac9975ebe9e3ba1752a948a/java-connectors/kafka-connect-azure-servicebus/src/test/java/io/lenses/streamreactor/connect/azure/servicebus/source/TaskToReceiverBridgeTest.java#L94)  I suppose I commented on the wrong method, but the amount of duplication here needs to be resolved.

Reply:
We're running in circles :) This is *my code* because I used your example.

But I know the answer: because we use stream there's no other way to deal with that exception, with `for` it'd be way more lenient.

Test has been updated